File: scripts/bundle.py
#!/usr/bin/env python3
import argparse
import hashlib
import json
import math
import re
import shutil
from datetime import datetime, timezone
from pathlib import Path


WORKLOADS = [
    "idle",
    "point-read-uniform",
    "point-read-skewed",
    "point-read-missing",
    "read-heavy",
    "balanced",
    "update-heavy",
    "range-scan",
    "sustained-ingest",
    "transaction-contention",
]
PATCH_DIRECTORY = Path(__file__).resolve().parent.parent / "patches" / "slatedb"


def arguments():
    parser = argparse.ArgumentParser(description="Assemble a SlateDB benchmark result bundle")
    parser.add_argument("--input", required=True, type=Path)
    parser.add_argument("--output", required=True, type=Path)
    parser.add_argument("--golden", required=True)
    parser.add_argument("--started-at", required=True)
    return parser.parse_args()


def read_result(path, task, golden):
    with path.open(encoding="utf-8") as file:
        result = json.load(file)
    if result.get("status") != "ok":
        raise ValueError(f"{path} does not contain a successful result")
    if result.get("task") != task:
        raise ValueError(f"{path} contains task {result.get('task')!r}, expected {task!r}")
    if result.get("golden_id") != golden:
        raise ValueError(f"{path} belongs to another golden data set")
    if result.get("recorded_interval_ns", 0) <= 0:
        raise ValueError(f"{path} has no recorded metrics")
    for field in ("environment", "application", "object_store", "process", "machine"):
        if not isinstance(result.get(field), dict):
            raise ValueError(f"{path} has no {field} metrics")
    return result


def read_golden(path, golden):
    with path.open(encoding="utf-8") as file:
        manifest = json.load(file)
    if manifest.get("status") != "ok":
        raise ValueError(f"{path} does not contain a successful golden manifest")
    if manifest.get("golden_id") != golden:
        raise ValueError(f"{path} belongs to another golden data set")
    if manifest.get("configuration", {}).get("task", {}).get("task") != "compaction":
        raise ValueError(f"{path} is not the final compacted golden manifest")
    for field in ("source", "environment", "checkpoint", "dataset"):
        if not isinstance(manifest.get(field), dict):
            raise ValueError(f"{path} has no {field}")
    return manifest


def sha256(path):
    return hashlib.sha256(path.read_bytes()).hexdigest()


def read_patches(directory=PATCH_DIRECTORY):
    return [
        {"name": path.name, "sha256": sha256(path)}
        for path in sorted(directory.glob("*.patch"))
        if path.is_file()
    ]


def read_series(path, result):
    if result.get("series", {}).get("file") != "series.json":
        raise ValueError(f"{path.parent / 'result.json'} has an invalid series reference")
    with path.open(encoding="utf-8") as file:
        series = json.load(file)
    required = {
        "rate_elapsed_ns",
        "rate_duration_ns",
        "latency_elapsed_ns",
        "latency_duration_ns",
        "resource_elapsed_ns",
        "resource_duration_ns",
        "application",
        "object_store",
        "process",
        "machine",
    }
    if set(series) != required:
        raise ValueError(f"{path} does not contain the expected workload series")
    if sha256(path) != result["series"].get("sha256"):
        raise ValueError(f"{path} does not match its result digest")
    return path


def read_transfer_capacity(path, scale, environment):
    with path.open(encoding="utf-8") as file:
        result = json.load(file)
    if result.get("status") != "ok":
        raise ValueError(f"{path} does not contain a successful transfer probe")
    if result.get("scale") != scale:
        raise ValueError(f"{path} used a different scale")
    for field in ("runner_type", "object_store", "endpoint", "region"):
        if result.get(field) != environment[field]:
            raise ValueError(f"{path} used a different {field.replace('_', ' ')}")
    validate_warp_transfer_capacity(path, result)
    return result


def validate_warp_transfer_capacity(path, result):
    if result.get("version") != 3:
        raise ValueError(f"{path} has unsupported transfer probe version")
    tool = result.get("tool")
    if (
        not isinstance(tool, dict)
        or tool.get("name") != "warp"
        or not isinstance(tool.get("version"), str)
        or not tool["version"]
    ):
        raise ValueError(f"{path} has invalid Warp tool metadata")
    benchmarks = result.get("benchmarks")
    if (
        not isinstance(benchmarks, list)
        or len(benchmarks) != 5
        or not all(isinstance(benchmark, dict) for benchmark in benchmarks)
    ):
        raise ValueError(f"{path} has invalid Warp benchmarks")
    expected = {
        "large-put": "PUT",
        "large-get": "GET",
        "small-put": "PUT",
        "small-get": "GET",
        "small-list": "LIST",
    }
    if {benchmark.get("name") for benchmark in benchmarks} != set(expected):
        raise ValueError(f"{path} has unexpected Warp benchmarks")
    for benchmark in benchmarks:
        name = benchmark["name"]
        if benchmark.get("operation") != expected[name]:
            raise ValueError(f"{path} has invalid {name} operation")
        for field in ("object_size_bytes", "concurrency", "duration_seconds"):
            value = benchmark.get(field)
            if not isinstance(value, int) or isinstance(value, bool) or value <= 0:
                raise ValueError(f"{path} has invalid {name} {field.replace('_', ' ')}")
        latency = benchmark.get("latency_ms")
        if latency is None:
            raise ValueError(f"{path} has no {name} latency")
        validate_warp_latency(path, name, latency)
        if benchmark.get("benchdata") != f"warp/{name}.csv.zst":
            raise ValueError(f"{path} has invalid {name} benchmark data")


def validate_warp_latency(path, name, latency):
    if not isinstance(latency, dict) or set(latency) not in (
        {"request"},
        {"request", "ttfb"},
    ):
        raise ValueError(f"{path} has invalid {name} latency")
    fields = {"average", "p50", "p90", "p99", "min", "max"}
    for kind, summary in latency.items():
        if not isinstance(summary, dict) or set(summary) != fields:
            raise ValueError(f"{path} has invalid {name} {kind} latency")
        for field, value in summary.items():
            if (
                not isinstance(value, (int, float))
                or isinstance(value, bool)
                or not math.isfinite(value)
                or value < 0
            ):
                raise ValueError(f"{path} has invalid {name} {kind} {field} latency")
        if not (
            summary["min"]
            <= summary["p50"]
            <= summary["p90"]
            <= summary["p99"]
            <= summary["max"]
        ):
            raise ValueError(f"{path} has unordered {name} {kind} latency")


def write_json(path, value):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(value, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def discover_workloads(directory):
    if not directory.is_dir():
        raise ValueError(f"{directory} does not contain workload results")
    discovered = {
        child.name
        for child in directory.iterdir()
        if child.is_dir() and (child / "result.json").is_file()
    }
    unknown = discovered.difference(WORKLOADS)
    if unknown:
        raise ValueError(f"{directory} contains unknown workloads: {', '.join(sorted(unknown))}")
    tasks = [task for task in WORKLOADS if task in discovered]
    if not tasks:
        raise ValueError(f"{directory} does not contain any workload results")
    return tasks


def validate_source_identities(workloads):
    if not workloads:
        raise ValueError("no workload results were provided")
    source = next(iter(workloads.values()))["source"]
    slate_commit = source["slate_commit"]
    runner_commit = source["runner_commit"]
    for task, result in workloads.items():
        if result["source"]["slate_commit"] != slate_commit:
            raise ValueError(f"{task} used a different SlateDB commit")
        if result["source"]["runner_commit"] != runner_commit:
            raise ValueError(f"{task} used a different benchmark runner commit")

    return source


def main():
    args = arguments()

    golden = read_golden(args.input / "golden.json", args.golden)
    workload_tasks = discover_workloads(args.input / "workload")
    workloads = {
        task: read_result(args.input / "workload" / task / "result.json", task, args.golden)
        for task in workload_tasks
    }
    workload_series = {
        task: read_series(args.input / "workload" / task / "series.json", workloads[task])
        for task in workload_tasks
    }

    source = validate_source_identities(workloads)
    version = source["slate_version"]
    first_workload = next(iter(workloads.values()))
    scale = first_workload["configuration"]["scale"]
    if not re.fullmatch(r"[A-Za-z0-9._-]{1,128}", version) or version in {".", ".."}:
        raise ValueError(f"invalid SlateDB version {version!r}")
    sessions = {result["session"] for result in workloads.values()}
    if len(sessions) != 1:
        raise ValueError("workload results belong to different sessions")
    run_id = next(iter(sessions))
    if not re.fullmatch(r"[A-Za-z0-9._-]{1,128}", run_id) or run_id in {".", ".."}:
        raise ValueError(f"invalid benchmark run ID {run_id!r}")
    for task, result in workloads.items():
        if result["configuration"]["scale"] != scale:
            raise ValueError(f"{task} used a different scale")
        if task == "sustained-ingest":
            if result["initial_state"]["kind"] != "empty":
                raise ValueError("sustained-ingest did not start empty")
        else:
            initial = result["initial_state"]
            checkpoint = golden["checkpoint"]
            if (
                initial["checkpoint_id"] != checkpoint["checkpoint_id"]
                or initial["manifest_id"] != checkpoint["manifest_id"]
                or initial["lsm_digest_sha256"] != checkpoint["lsm_digest_sha256"]
            ):
                raise ValueError(f"{task} did not start from the golden checkpoint")
    if golden["configuration"]["scale"] != scale:
        raise ValueError("golden manifest used a different scale")

    transfer_capacity_path = args.input / "transfer-capacity" / "result.json"
    transfer_capacity = (
        read_transfer_capacity(
            transfer_capacity_path,
            scale,
            first_workload["environment"],
        )
        if transfer_capacity_path.is_file()
        else None
    )

    destination = args.output / version / run_id
    if destination.exists():
        shutil.rmtree(destination)
    golden_target = destination / "golden.json"
    write_json(golden_target, golden)
    checksums = {"golden.json": sha256(golden_target)}
    configurations = {"golden": golden["configuration"]}
    for task, result in workloads.items():
        relative = Path("workload") / task / "result.json"
        target = destination / relative
        write_json(target, result)
        checksums[relative.as_posix()] = sha256(target)
        configurations[task] = result["configuration"]
        series_relative = Path("workload") / task / "series.json"
        series_target = destination / series_relative
        series_target.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(workload_series[task], series_target)
        checksums[series_relative.as_posix()] = sha256(series_target)

    manifest = {
        "status": "ok",
        "run_id": run_id,
        "golden_id": args.golden,
        "started_at": args.started_at,
        "finished_at": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "patches": read_patches(),
        "source": source,
        "golden_runner_commit": golden["source"]["runner_commit"],
        "resolved_configuration": configurations,
        "max_parallel": len(workloads),
        "results": checksums,
    }
    if transfer_capacity is not None:
        manifest["transfer_capacity"] = transfer_capacity
    write_json(destination / "run.json", manifest)
    print(destination)


if __name__ == "__main__":
    main()

File: src/bundle.rs
//! Assembly of validated workload outputs into a publishable run directory.

use crate::config::Task;
use crate::model::{
    AppliedPatch, GoldenManifest, RunManifest, SourceIdentity, WorkloadResult, WorkloadSeries,
};
use crate::validation::{
    validate_golden_manifest, validate_identifier, validate_run_manifest, validate_workload_result,
    validate_workload_series,
};
use anyhow::{ensure, Context, Result};
use chrono::Utc;
use sha2::{Digest, Sha256};
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};

/// Inputs used to assemble a complete benchmark run bundle.
pub struct BundleArgs {
    /// Directory containing `golden.json` and per-workload result directories.
    pub input: PathBuf,
    /// Root under which the versioned run directory will be created.
    pub output: PathBuf,
    /// Identifier of the golden dataset used by the run.
    pub golden: String,
    /// RFC 3339 timestamp captured when the run began.
    pub started_at: String,
    /// Directory containing SlateDB patches applied to the run.
    pub patches: PathBuf,
}

/// Validates the supplied artifacts and writes a self-contained run bundle.
///
/// The returned path is `<output>/<slatedb-version>/<session>`.
pub fn bundle(args: BundleArgs) -> Result<PathBuf> {
    validate_identifier(&args.golden, "golden ID")?;
    let golden_path = args.input.join("golden.json");
    let golden: GoldenManifest = read_json(&golden_path)?;
    validate_golden_manifest(&golden)
        .with_context(|| format!("validating {}", golden_path.display()))?;
    ensure!(
        golden.golden_id == args.golden,
        "{} belongs to another golden data set",
        golden_path.display()
    );
    ensure!(
        golden.configuration.task.task == Task::Compaction,
        "{} is not the final compacted golden manifest",
        golden_path.display()
    );

    let tasks = discover_workloads(&args.input.join("workload"))?;
    let mut workloads = BTreeMap::new();
    let mut series = BTreeMap::new();
    for task in tasks {
        let directory = args.input.join("workload").join(task.as_str());
        let result_path = directory.join("result.json");
        let result: WorkloadResult = read_json(&result_path)?;
        validate_workload_result(&result)
            .with_context(|| format!("validating {}", result_path.display()))?;
        ensure!(
            result.task == task,
            "{} contains task {}, expected {}",
            result_path.display(),
            result.task,
            task
        );
        ensure!(
            result.golden_id == args.golden,
            "{} belongs to another golden data set",
            result_path.display()
        );
        let series_path = directory.join("series.json");
        ensure!(
            sha256_file(&series_path)? == result.series.sha256,
            "{} does not match its result digest",
            series_path.display()
        );
        let workload_series: WorkloadSeries = read_json(&series_path)?;
        validate_workload_series(&result, &workload_series)
            .with_context(|| format!("validating {}", series_path.display()))?;
        workloads.insert(task, result);
        series.insert(task, series_path);
    }

    let first = workloads
        .values()
        .next()
        .context("no workload results were provided")?;
    let source = validate_source_identities(&workloads)?;
    validate_identifier(&source.slate_version, "SlateDB version")?;
    validate_identifier(&first.session, "run ID")?;
    let scale = first.configuration.scale;
    for (task, result) in &workloads {
        ensure!(
            result.session == first.session,
            "workload results belong to different sessions"
        );
        ensure!(
            result.configuration.scale == scale,
            "{task} used a different scale"
        );
        if *task == Task::SustainedIngest {
            ensure!(
                result.initial_state.kind == "empty",
                "sustained-ingest did not start empty"
            );
        } else {
            ensure!(
                result.initial_state.checkpoint_id.as_deref()
                    == Some(golden.checkpoint.checkpoint_id.as_str())
                    && result.initial_state.manifest_id == Some(golden.checkpoint.manifest_id)
                    && result.initial_state.lsm_digest_sha256
                        == golden.checkpoint.lsm_digest_sha256,
                "{task} did not start from the golden checkpoint"
            );
        }
    }
    ensure!(
        golden.configuration.scale == scale,
        "golden manifest used a different scale"
    );

    let destination = args.output.join(&source.slate_version).join(&first.session);
    if destination.exists() {
        fs::remove_dir_all(&destination)
            .with_context(|| format!("removing {}", destination.display()))?;
    }
    fs::create_dir_all(&destination)
        .with_context(|| format!("creating {}", destination.display()))?;

    let golden_target = destination.join("golden.json");
    write_json(&golden_target, &golden)?;
    let mut results = BTreeMap::from([("golden.json".to_string(), sha256_file(&golden_target)?)]);
    let mut resolved_configuration =
        BTreeMap::from([("golden".to_string(), golden.configuration.clone())]);
    for (task, result) in &workloads {
        let relative = format!("workload/{}/result.json", task.as_str());
        let target = destination.join(&relative);
        write_json(&target, result)?;
        results.insert(relative, sha256_file(&target)?);
        resolved_configuration.insert(task.as_str().to_string(), result.configuration.clone());

        let relative = format!("workload/{}/series.json", task.as_str());
        let target = destination.join(&relative);
        fs::copy(&series[task], &target).with_context(|| {
            format!("copying {} to {}", series[task].display(), target.display())
        })?;
        results.insert(relative, sha256_file(&target)?);
    }

    let manifest = RunManifest {
        status: "ok".to_string(),
        run_id: first.session.clone(),
        golden_id: args.golden,
        started_at: args.started_at,
        finished_at: Utc::now().to_rfc3339(),
        patches: read_patches(&args.patches)?,
        source,
        golden_runner_commit: golden.source.runner_commit.clone(),
        resolved_configuration,
        max_parallel: workloads.len(),
        results,
    };
    validate_run_manifest(&manifest)?;
    write_json(&destination.join("run.json"), &manifest)?;
    Ok(destination)
}

fn discover_workloads(directory: &Path) -> Result<Vec<Task>> {
    ensure!(
        directory.is_dir(),
        "{} does not contain workload results",
        directory.display()
    );
    let allowed = Task::workloads()
        .map(|task| (task.as_str(), task))
        .collect::<BTreeMap<_, _>>();
    let mut discovered = BTreeSet::new();
    for entry in
        fs::read_dir(directory).with_context(|| format!("reading {}", directory.display()))?
    {
        let entry = entry?;
        if entry.file_type()?.is_dir() && entry.path().join("result.json").is_file() {
            let name = entry.file_name();
            let name = name.to_string_lossy();
            let task = allowed.get(name.as_ref()).copied().with_context(|| {
                format!("{} contains unknown workload {name}", directory.display())
            })?;
            discovered.insert(task);
        }
    }
    ensure!(
        !discovered.is_empty(),
        "{} does not contain any workload results",
        directory.display()
    );
    Ok(Task::workloads()
        .filter(|task| discovered.contains(task))
        .collect())
}

fn validate_source_identities(
    workloads: &BTreeMap<Task, WorkloadResult>,
) -> Result<SourceIdentity> {
    let first = workloads
        .values()
        .next()
        .context("no workload results were provided")?;
    for (task, result) in workloads {
        ensure!(
            result.source.slate_commit == first.source.slate_commit,
            "{task} used a different SlateDB commit"
        );
        ensure!(
            result.source.runner_commit == first.source.runner_commit,
            "{task} used a different benchmark runner commit"
        );
    }
    Ok(first.source.clone())
}

fn read_patches(directory: &Path) -> Result<Vec<AppliedPatch>> {
    if !directory.exists() {
        return Ok(Vec::new());
    }
    let mut paths = fs::read_dir(directory)
        .with_context(|| format!("reading {}", directory.display()))?
        .filter_map(|entry| entry.ok().map(|entry| entry.path()))
        .filter(|path| path.is_file() && path.extension().is_some_and(|value| value == "patch"))
        .collect::<Vec<_>>();
    paths.sort();
    paths
        .into_iter()
        .map(|path| {
            Ok(AppliedPatch {
                name: path
                    .file_name()
                    .context("patch has no file name")?
                    .to_string_lossy()
                    .into_owned(),
                sha256: sha256_file(&path)?,
            })
        })
        .collect()
}

/// Computes the lowercase SHA-256 digest of a file.
pub fn sha256_file(path: &Path) -> Result<String> {
    let contents = fs::read(path).with_context(|| format!("reading {}", path.display()))?;
    Ok(format!("{:x}", Sha256::digest(contents)))
}

fn read_json<T: serde::de::DeserializeOwned>(path: &Path) -> Result<T> {
    let file = fs::File::open(path).with_context(|| format!("opening {}", path.display()))?;
    serde_json::from_reader(file).with_context(|| format!("parsing {}", path.display()))
}

fn write_json<T: serde::Serialize>(path: &Path, value: &T) -> Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).with_context(|| format!("creating {}", parent.display()))?;
    }
    let mut contents = serde_json::to_vec_pretty(value)?;
    contents.push(b'\n');
    fs::write(path, contents).with_context(|| format!("writing {}", path.display()))
}

#[cfg(test)]
mod tests {
    use super::{discover_workloads, read_patches};
    use crate::config::Task;
    use std::fs;

    #[test]
    fn discovers_a_workload_subset_in_catalog_order() {
        let temporary = tempfile::tempdir().expect("create temporary directory");
        for name in ["sustained-ingest", "balanced"] {
            let path = temporary.path().join(name);
            fs::create_dir(&path).expect("create workload directory");
            fs::write(path.join("result.json"), "{}").expect("write workload result");
        }
        assert_eq!(
            discover_workloads(temporary.path()).expect("discover workloads"),
            vec![Task::Balanced, Task::SustainedIngest]
        );
    }

    #[test]
    fn records_patches_in_filename_order() {
        let temporary = tempfile::tempdir().expect("create temporary directory");
        fs::write(temporary.path().join("0002.patch"), "second").expect("write second patch");
        fs::write(temporary.path().join("0001.patch"), "first").expect("write first patch");
        fs::write(temporary.path().join("README.md"), "ignored").expect("write ignored file");
        let patches = read_patches(temporary.path()).expect("read patches");
        assert_eq!(
            patches
                .iter()
                .map(|patch| patch.name.as_str())
                .collect::<Vec<_>>(),
            ["0001.patch", "0002.patch"]
        );
    }
}
